pkg/cmd/ssh/command.go

```diff
@@ -0,0 +1,801 @@
+			return fmt.Errorf("failed to output node table: %w", err)
+		}
+
+		fmt.Fprintln(o.IOStreams.Out, "")
+	}
+
+	fmt.Fprintln(o.IOStreams.Out, "Connect to shoot nodes by using the bastion as a proxy/jump host, for example:")
+	fmt.Fprintln(o.IOStreams.Out, "")
+	fmt.Fprintln(o.IOStreams.Out, connectCmd)
+	fmt.Fprintln(o.IOStreams.Out, "")
+
+	fmt.Fprintln(o.IOStreams.Out, "Press Ctrl-C to stop gardenctl, after which the bastion will be removed.")
+
+	<-signalChan
+
+	return nil
+}
+
+func isNodeReady(node corev1.Node) bool {
+	for _, cond := range node.Status.Conditions {
+		if cond.Type == corev1.NodeReady {
+			return cond.Status == corev1.ConditionTrue
+		}
+	}
+
+	return false
+}
+
+func sshCommandLine(o *Options, bastionAddr string, nodePrivateKeyFiles []string, nodeName string) string {
+	proxyPrivateKeyFlag := ""
+	if o.SSHPrivateKeyFile != "" {
+		proxyPrivateKeyFlag = fmt.Sprintf(" -o IdentitiesOnly=yes -i %s", o.SSHPrivateKeyFile)
+	}
+
+	proxyCmd := fmt.Sprintf(
+		"ssh -W%%h:%%p -o StrictHostKeyChecking=no%s %s@%s",
+		proxyPrivateKeyFlag,
+		SSHBastionUsername,
+		bastionAddr,
+	)
+
+	identities := []string{}
+	for _, filename := range nodePrivateKeyFiles {
+		identities = append(identities, fmt.Sprintf("-i %s", filename))
+	}
+
+	connectCmd := fmt.Sprintf(
+		`ssh -o "StrictHostKeyChecking=no" -o "IdentitiesOnly=yes" %s -o "ProxyCommand=%s" %s@%s`,
+		strings.Join(identities, " "),
+		proxyCmd,
+		SSHNodeUsername,
+		nodeName,
+	)
+
+	return connectCmd
+}
+
+func keepBastionAlive(ctx context.Context, gardenClient client.Client, bastion *operationsv1alpha1.Bastion, stderr io.Writer) {
+	ticker := time.NewTicker(keepAliveInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-ticker.C:
+			// re-fetch current bastion
+			key := types.NamespacedName{Name: bastion.Name, Namespace: bastion.Namespace}
+
+			// reset annotations so that we fetch the actual current state
+			bastion.Annotations = map[string]string{}
+
+			if err := gardenClient.Get(ctx, key, bastion); err != nil {
+				fmt.Fprintf(stderr, "Failed to keep bastion alive: %v\n", err)
+			}
+
+			// add the keepalive annotation
+			oldBastion := bastion.DeepCopy()
+			bastion.Annotations[v1beta1constants.GardenerOperation] = v1beta1constants.GardenerOperationKeepalive
+
+			if err := gardenClient.Patch(ctx, bastion, client.MergeFrom(oldBastion)); err != nil {
+				fmt.Fprintf(stderr, "Failed to keep bastion alive: %v\n", err)
+			}
+		}
+	}
+}
+
+func getShootNodePrivateKeys(ctx context.Context, gardenClient client.Client, shoot *gardencorev1beta1.Shoot) ([][]byte, error) {
+	secret := &corev1.Secret{}
+	key := types.NamespacedName{
+		// TODO: use ShootProjectSecretSuffixSSHKeypair once Gardener supports ctrl-runtime 0.9
+		Name:      fmt.Sprintf("%s.ssh-keypair", shoot.Name),
+		Namespace: shoot.Namespace,
+	}
+
+	if err := gardenClient.Get(ctx, key, secret); err != nil {
+		return nil, err
+	}
```

Comment:
there might be a time window where this secret is not there, see https://github.com/gardener/gardener/blob/f54564c83d8dee91eb46bc726119a8d72d578810/pkg/operation/botanist/secrets.go#L287-L289

I think in this case we could also ignore not found errors. 

However if `keys` slice is empty (both secrets were not found) it should be considered as error, of course

Reply:
I don't quite understand. If the "primary" SSH key isn't there, the secondary (".old") will also not be there. So if we ignore the NotFound error on the primary, the secondary will also fail and we will run into the "slice is empty" situation, i.e. an error.

Or are you referring to a different timeframe than shortly after a Shoot has been created?

Reply:
during `rotateSSHKeypairSecrets` (see link above) the current secret get's deleted and recreated later, however at that point in time the old key should exist

```go
	// delete the current key so it gets re-generated later
	if err := b.K8sSeedClient.Client().Delete(ctx, currentSecret); client.IgnoreNotFound(err) != nil {
		return fmt.Errorf("failed to delete current SSH keypair: %v", err)
	}
```